src/rt/ds/stack.h

```diff
@@ -7,128 +7,302 @@
+    /// correct to call this if pop_is_fast just returned
+    /// true.
+    ALWAYSINLINE T* pop_fast()
+    {
+      assert(pop_is_fast());
+      auto item = peek();
       index--;
-      T item = block->data[index];
+      return item;
+    }
 
-      if (index == 0)
-        pop_slow_path();
+    /// Call this to pop an element from the stack.  Only
+    /// correct to call this if pop_is_fast just returned
+    /// false.  This returns a pair of the popped element
+    /// and the block that the client must dispose of.
+    std::pair<T*, Block*> pop_slow()
+    {
+      auto item = peek();
+      T** prev_index = get_block(index)->prev;
+      auto dealloc = get_block(index);
+      index = prev_index;
+      return {item, dealloc};
+    }
 
-      return item;
+    /// Call this to determine if push can proceed without
+    /// a new block.
+    ALWAYSINLINE bool push_is_fast()
+    {
+      return is_not_full(index);
     }
 
-    ALWAYSINLINE void push(T item)
+    /// Call this to push an element onto the stack.  Only
+    /// correct to call this if push_is_fast just returned
+    /// true.
+    ALWAYSINLINE void push_fast(T* item)
     {
-      if (index < STACK_COUNT)
-      {
-        block->data[index] = item;
-        index++;
-      }
-      else
-      {
-        push_slow_path(item);
-      }
+      assert(push_is_fast());
+      index++;
+      *index = item;
     }
 
-    template<void apply(T t)>
+    /// Call this to push an element onto the stack.  Only
+    /// correct to call this if push_is_fast just returned
+    /// false.  It needs to be provided a new block of memory
+    /// for the stack to use.
+    void push_slow(T* item, Block* block)
+    {
+      assert(!push_is_fast());
+
+      T** iter = (T**)block;
+      assert(is_empty(iter));
+      auto next = get_block(iter);
+
+      assert(index != (T**)&null_block);
+      next->prev = index;
+      index = &(next->data[0]);
+      next->data[0] = item;
+    }
+
+    /// For all elements of the stack
+    template<void apply(T* t)>
     void forall()
     {
-      Block* curr = block;
-      size_t i = index;
+      T* curr = index;
 
-      while (curr != nullptr)
+      while (curr != null_index)
       {
         do
         {
-          i--;
-          apply(curr->data[i]);
-        } while (i > 0);
+          apply(*curr);
+          curr--;
+        } while (is_empty(curr));
 
-        curr = curr->prev;
-        i = STACK_COUNT;
+        curr = get_block(curr)->prev;
       }
     }
+  };
 
-  private:
-    void pop_slow_path()
+  /**
+   * This class uses the block structured stack with extra fields
+   * for the allocator, and a backup block, so that the common case
+   * of 0-1 elements can be fast, and any other block boundrary case.
+   */
+  template<class T, class Alloc>
+  class Stack
+  {
+    using Block = typename StackBase<T, Alloc>::Block;
+    StackBase<T, Alloc> stack;
+    Alloc* alloc;
```

Comment:
We should add a note to the docstring that this is not expected to change threads